### ISSUE TITLE
use default build-in transport in reliable mode, the last one packet or last several packets can not be received by subscriber

### ISSUE THREAD

**Issue by yuzu-ogura** [eProsima/Fast-DDS#5280]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

in reliable mode, all packet maybe received by subscriber

### Current behavior

 the last one packet or last several packets can not be received by subscriber

### Steps to reproduce

// high frequency send data
while(i < 1000) {
    datawriter->write(userdata);
}

### Fast DDS version/commit

v2.6.9

### Platform/Architecture

Ubuntu Focal 20.04 amd64

### Transport layer

Default configuration, UDPv4 & SHM

**Comment by yuzu-ogura** [comment]
of course @JesusPoderoso .
datawriterqos is like below:
```
    DataWriterQos writerQos(DATAWRITER_QOS_DEFAULT);
    writerQos.resource_limits().max_samples = 1000;
    writerQos.history().kind = KEEP_ALL_HISTORY_QOS;
    writerQos.publish_mode().kind = ASYNCHRONOUS_PUBLISH_MODE;
    writerQos.reliability().kind = RELIABLE_RELIABILITY_QOS;
    writerQos.data_sharing().off();
    writerQos.endpoint().history_memory_policy = eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
```
datareaderqos is like below:
```
    eprosima::fastdds::dds::DataReaderQos readerQos(DATAREADER_QOS_DEFAULT);
    readerQos.resource_limits().max_samples = 1000;
    readerQos.history().kind = KEEP_ALL_HISTORY_QOS;
    readerQos.endpoint().history_memory_policy = eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
    readerQos.data_sharing().off();
    readerQos.reliability().kind = RELIABLE_RELIABILITY_QOS;
    readerQos.durability().kind = TRANSIENT_LOCAL_DURABILITY_QOS;
```
other qos all apply the default qos
main.cpp like below:
```
while(index < 999) {
        HelloWorld hello;
        hello.index(++index);
        datawriter->write(&hello)
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
```

idl like below:
```
struct HelloWorld
{
    unsigned long index;
    sequence<octet> points;
};
```

**Comment by JesusPoderoso** [comment]
@yuzu-ogura are the endpoints running in the same process?
In that case, do you wait for the subscriber to receive all samples before killing the publisher and ending the app execution?

**Comment by JesusPoderoso** [comment]
Using the [wait-for-acknowledgments method](https://fast-dds.docs.eprosima.com/en/latest/fastdds/api_reference/dds_pim/publisher/datawriter.html#_CPPv4N8eprosima7fastdds3dds10DataWriter24wait_for_acknowledgmentsERKN7fastdds3dds10Duration_tE) forces the publisher to wait until the sample has been acknowledged. You may consider using the method on the last sample to force that @yuzu-ogura.

You may also consider checking the `write` call output just in case the write is not being performed (due to a timeout).
To solve that, you should [increase the `max_blocking_time`](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/realtime/blocking.html#non-blocking-calls) reliability QoS field.

**Comment by yuzu-ogura** [comment]
> @yuzu-ogura are the endpoints running in the same process? In that case, do you wait for the subscriber to receive all samples before killing the publisher and ending the app execution?

yes. pub and sub are run separately on different host. And pub will only exit when all the data had been sent . The sub will always sleep in main thread to avoid exiting.

**Comment by yuzu-ogura** [comment]
> You may also consider checking the `write` call output just in case the write is not being performed (due to a timeout).
> To solve that, you should [increase the `max_blocking_time`](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/realtime/blocking.html#non-blocking-calls) reliability QoS field.

so, if `bool write()` return true, it doesn't mean data send successfully?

**Comment by JesusPoderoso** [comment]
> so, if bool write() return true, it doesn't mean data send successfully?

The `write` return value was not considered in the code snippet provided. It returns `true` if successfully sent, but it can also return `false`. In those cases, you were not considering that the `write` call failed.

**Comment by JesusPoderoso** [comment]
> pub will only exit when all the data had been sent 

In such a reliable case scenario, you should wait until all data has been received too.

**Comment by yuzu-ogura** [comment]
> The `write` return value was not considered in the code snippet provided. It returns `true` if successfully sent, but it can also return `false`. In those cases, you were not considering that the `write` call failed.

I remove other non-necessary code include check reture value fromcode snippet . And here is pub && sub log:
pub:
![image](https://github.com/user-attachments/assets/f09f0117-3b21-4f33-bca5-ed8f2a30597a)

sub:
![image](https://github.com/user-attachments/assets/24e1bb21-7672-4082-9b64-faafa9d5321e)

**Comment by yuzu-ogura** [comment]
> > pub will only exit when all the data had been sent
> 
> In such a reliable case scenario, you should wait until all data has been received too.

This approach will block until sub really receive data or timeout. However, it may reduce the sending rate ?

**Comment by yuzu-ogura** [comment]
@JesusPoderoso Sorry, I realized it was my fault .Thanks for your enthusiastically reply . Just now I found another scene that call `write` method before pub && sub matched. the `write` method still return true. how could I solve it ?

**Comment by JesusPoderoso** [comment]
> This approach will block until sub really receive data or timeout.

Yes

> However, it may reduce the sending rate?

No, because you only wait the ack of the sent samples after all samples have been sent.

``` cpp
    while(index < 999) {
        HelloWorld hello;
        hello.index(++index);
        if (!datawriter->write(&hello)){
           std::cout << "Could not send message with  index " << index << std::endl;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
    datawriter->wait_for_acknowledgments(eprosima::fastrtps::Duration_t(1, 0));  // wait 1 sec 
```

**Comment by JesusPoderoso** [comment]
>  Just now I found another scene that call write method before pub && sub matched. the write method still return true. how could I solve it ?

I would recommend you @yuzu-ogura to follow the best-practices structure of the refactored [examples](https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/hello_world) in Fast DDS v3. In particular, the mechanism to determine that [endpoints have already matched](https://github.com/eProsima/Fast-DDS/blob/master/examples/cpp/hello_world/PublisherApp.cpp#L152) before start sending messages.

**Comment by yuzu-ogura** [comment]
OK, Thanks for your enthusiastically reply.
